sources.sh

```diff
@@ -254,15 +270,26 @@ jq <<<"$sources" --argjson pins "$externalPinsJson" '
 	# TODO a lot of this could be removed/parsed during the above reduce, since it has to parse things in build order anyhow
 	# TODO actually, instead, this bit should be a totally separate script so the use case of "combine sources.json files together" works better 👀
 	| (
-		# TODO make this faster, this reduce takes the longest time now
-		reduce to_entries[] as $e ({};
-			$e.key as $sourceId
-			| .[ $e.value.arches[] | .tags[], .archTags[] ] |= (
-				.[$e.value.arches | keys[]] |= (
-					. + [$sourceId] | unique_unsorted
-				)
+		# creating a lookup of .[tag][arch] to a list of sourceIds
+		[
+			.[] as $e
+			| ( $e.arches[] | .tags[], .archTags[] )
+			| {
+				key: .,
+				value: {
+					($e.arches | keys[]): [$e.sourceId]
```

Comment:
This is _technically_ looping over arches twice and creating combinations that don't exist (just like the previous implementation), but it's fine.

Reply:
So... yeah, moving this to before the reduce in order to get rid of that duplication (along with your meld improvements), makes it even faster!!!

```console
$ time ../meta-scripts/sources.sh --cache-file=sources-copy.json > sources.json

real    0m5.095s
user    0m4.570s
sys     0m0.277s
```